src/crypto/clu_decrypt.c

```diff
@@ -0,0 +1,223 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+#include <wolfclu/clu_header_main.h>
+#include <wolfclu/clu_log.h>
+#include <wolfclu/clu_optargs.h>
+#include <wolfclu/genkey/clu_genkey.h>
+
+#define MAX_LEN             1024
+
+/* Used for algorithms that do not have an EVP type
+ * return WOLFCLU_SUCCESS on success
+ */
+int wolfCLU_decrypt(int alg, char* mode, byte* pwdKey, byte* key, int size,
+        char* in, char* out, byte* iv, int block, int keyType)
+{
+#ifdef HAVE_CAMELLIA
+    Camellia camellia;                  /* camellia declaration */
+#endif
+
+    XFILE  inFile;                      /* input file */
+    XFILE  outFile;                     /* output file */
+
+    WC_RNG     rng;                     /* random number generator */
+    byte*   input  = NULL;              /* input buffer */
+    byte*   output = NULL;              /* output buffer */
+    byte    salt[SALT_SIZE] = {0};      /* salt variable */
+
+    int     currLoopFlag = 1;           /* flag to track the loop */
+    int     lastLoopFlag = 0;           /* flag for last loop */
+    int     ret          = 0;           /* return variable */
+    int     keyVerify    = 0;           /* verify the key is set */
+    int     i            = 0;           /* loop variable */
+    int     pad          = 0;           /* the length to pad */
+    int     length;                     /* length of message */
+    int     tempMax = MAX_LEN;              /* equal to MAX_LEN until feof */
+    int     sbSize = SALT_SIZE + block; /* size of salt and iv together */
+
+    /* opens input file */
+    inFile = XFOPEN(in, "rb");
+    if (inFile == NULL) {
+        WOLFCLU_LOG(WOLFCLU_L0, "Input file does not exist.");
+        return DECRYPT_ERROR;
+    }
+    /* opens output file */
+
+    if ((outFile = XFOPEN(out, "wb")) == NULL) {
+        WOLFCLU_LOG(WOLFCLU_L0, "Error creating output file.");
+        XFCLOSE(inFile);
+        return DECRYPT_ERROR;
+    }
+
+    /* find end of file for length */
+    XFSEEK(inFile, 0, SEEK_END);
+    length = (int)XFTELL(inFile);
+    XFSEEK(inFile, 0, SEEK_SET);
+
+    /* if there is a remainder,
+     * round up else no round
+     */
+    if (length % MAX_LEN > 0) {
+        lastLoopFlag = (length/MAX_LEN) + 1;
+    }
+    else {
+        lastLoopFlag =  length/MAX_LEN;
+    }
+
+    input = (byte*) XMALLOC(MAX_LEN, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
+    if (input == NULL) {
+        ret = MEMORY_E;
+    }
+
+    if (ret == 0) {
+        output = (byte*) XMALLOC(MAX_LEN, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
+        if (output == NULL) {
+            ret = MEMORY_E;
+        }
+    }
+
+    if (ret == 0) {
+        ret = wc_InitRng(&rng);
+    }
+
+    /* reads from inFile and writes whatever
+     * is there to the input buffer
+     */
+    while (length > 0 && ret == 0) {
+
+        /* On first loop only read in salt and iv */
```

Comment:
I may have said this before, but we should just do this before the loop if it only happens on the first iteration.